### PR TITLE
feat: Machine learning based Ambiguity Solver

### PR DESCRIPTION
**Pull request by Corentin-Allaire**
This MR depends on #1868.

Implement a machine learning based ambiguity solver. If Acts is compiled with the `ACTS_PLUGIN_ONNX` option, it can be used in the odd full chain by simply adding the `--MLSolver` argument.

In addition to the implementation of the solver, this PR adds an option for the regular solver to only work on track with a certain amount of hits (to obtain fairer comparison). 

This also add some small change to the Csv multitrajectory writing, it now also write the list of measurements associated with each track.

Finally, the python script to train the network, test the network and compute its performances are made available with the onnx file of the model. 

I haven't written the documentation for this solver, yet it will be part of a future MR after this gets merged in.

### CODE REVIEW

**Comment by asalzburger** [comment]
This is on top of the batch input for ONNX #1868 ?

**Comment by Corentin-Allaire** [comment]
> This is on top of the batch input for ONNX #1868 ?

Yes, the batch input caused a factor 7 speed up so I thought it should go in first. But I guess it is close to be merged

**Comment by asalzburger** [comment]
Alright, I have seen that you had addressed @benjaminhuth s comments, so I merged #1868

**Comment by Corentin-Allaire** [comment]
The jobs that don't pass the CI are related to the fact I have change the default solver to only consider track with at least 7 hits. I can put the threshold to 0 but honestly I would be concern by the quality of a track with < 7 hits.
If anybody has an opinion on this...

**Comment by asalzburger** [comment]
> e CI are rel

In that case, the PR should include the updated reference, what do you think? @paulgessinger

File: Examples/Python/src/TrackFinding.cpp
Comment (Corentin-Allaire) [reply Examples/Python/src/TrackFinding.cpp:13]: I am clearly not a cmake expert, would that be better ? If yes could we implement that in a later PR ? I don't really have the time to dive in the cmake doc right now...

Comment (andiwand) [reply Examples/Python/src/TrackFinding.cpp:13]: sure no worries I can do that after we got this in

File: Examples/Scripts/Python/full_chain_odd.py
Comment (andiwand) [inline Examples/Scripts/Python/full_chain_odd.py:169]: I guess because of that @paulgessinger

**Comment by Corentin-Allaire** [comment]
> Is it obvious why the non-ambi outputs also change @Corentin-Allaire?

I the odd-full-chain the vertexing uses the output of the Solver as an input. If you remove some tracks (the one with < 7 measurements) then it will affect the vertexing. What was a bit surprising is that it appear to improve the vertexing performance.

File: Examples/Scripts/Python/full_chain_odd.py
Comment (andiwand) [reply Examples/Scripts/Python/full_chain_odd.py:169]: oh yeah sorry. but I guess it is the one below where it also changed from 6 to 7 right?

Comment (andiwand) [reply Examples/Scripts/Python/full_chain_odd.py:169]: here https://github.com/acts-project/acts/pull/1877/files#diff-218e044dd033aa9356a7aa4785f709eef7f2a72f1b7cb7e3db5a95006efc3b9bL170-L178

Comment (andiwand) [reply Examples/Scripts/Python/full_chain_odd.py:169]: but I guess this is also only for performance writing

Comment (Corentin-Allaire) [reply Examples/Scripts/Python/full_chain_odd.py:169]: Yes, what is changing the physmon is this line : https://github.com/acts-project/acts/pull/1877/files#diff-311aa2ecef0d9d175b50ee63b8107c6ef3ef32fe6d3276d1ad462201f4b16891R127-R131

File: Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/AmbiguityResolutionAlgorithm.hpp
Comment (andiwand) [inline Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/AmbiguityResolutionAlgorithm.hpp:42]: this is the change which causes physmon to create a diff, right @Corentin-Allaire ? cc @paulgessinger

Comment (Corentin-Allaire) [reply Examples/Algorithms/TrackFinding/include/ActsExamples/TrackFinding/AmbiguityResolutionAlgorithm.hpp:42]: Yes ! As I mention before In my opinion with the geometry of the ODD a track with < 7 measurement is very unlikely to be a good track

File: Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp
Comment (asalzburger) [inline Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:187]: I really enjoyed seeing this and particularly your talk in today's meeting.

I have a bit of a design question:
- suppose we wanted to use this ML solver in ATLAS or sPHENIX wants to use it.
- can we encapsulate this code into the `Core` or a new `Core/ML` in order to make this algorithm not more than an empty shell that does input/output, but then pushes the prepared data into a `Core/ML/DuplicateSolver` ?

Comment (benjaminhuth) [reply Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:187]: I'm not sure if we should have ONNX as a dependency of the `Core` module? If we push this code to the Onnx-Plugin, wouldn't it also be usable in ATLAS / sPHENIX?

Comment (Corentin-Allaire) [reply Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:187]: I don't have a strong opinion either way, in core we could have only `Core/ML` depend on Onnx. What I dislike a bit about the Plugin in that case is that if we start to have more ML base algorithm they are all going to end up in the Onnx plugin and if we start to combine multiple ML library this will also cause some some issue. We could also have a ML Plugin but at that point what is the difference between this and a `Core/ML` directory ?

Comment (Corentin-Allaire) [reply Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:187]: But I would argue that maybe this should be a general discussion in one of our weekly meeting on how we want to integrate ML development in Acts. And maybe for now we should just add this as is in Example ?

Comment (paulgessinger) [reply Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:187]: For other plugins we have a split between a plugin and an examples "plugin". You need both to run it. For the python bindings, we can optionally include the ONNX examples in the python module.

Comment (Corentin-Allaire) [reply Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:187]: So at this point what should we do ? Do We keep the code as is for now ? Do I move it to the Onnx Plugin or another one ? Or even to Core ? 

I think it would be in everyone interest to have this merge so we can add this to the test/CI.

Comment (benjaminhuth) [reply Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:187]: I would vote for moving the parts that do not depend on the examples EDM to the plugin for now. Then it can be used without the Examples framework. However, one problem I currently see the usage of the `ActsExamples::TrajectoriesContainer`. Not sure if this could be easily replaced by the `Acts::TrackContainer` component? @paulgessinger 

If that requires some more shuffling around I would just leave it as is and refactor this at a later point.

Comment (Corentin-Allaire) [reply Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:187]: I am not too sure either how easy it would be to replace the `ActsExamples::TrajectoriesContainer` by `Acts::TrackContainer` ...

Comment (benjaminhuth) [reply Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:187]: Okay then I would say let's refactor this later and concentrate on the CI for now

Comment (andiwand) [reply Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:187]: I think that is a story for another day as it would also effect a lot of other components. to me it makes sense to put this in for now as is. otherwise I fear we just outdate and stale it

Comment (Corentin-Allaire) [reply Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:187]: Works for me. As soon as we merge this I can start working on a PR to add it to the CI.
Now the question is can someone review it ?

Comment (Corentin-Allaire) [reply Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:187]: I probably also need to update the physmon again...

File: Examples/Scripts/Python/full_chain_odd.py
Comment (andiwand) [inline Examples/Scripts/Python/full_chain_odd.py:177]: is this relevant for training? otherwise I would just drop it since we are writing out the root files as well

same fore the ambiguity below

File: Examples/Algorithms/TrackFinding/src/AmbiguityResolutionAlgorithm.cpp
Comment (andiwand) [inline Examples/Algorithms/TrackFinding/src/AmbiguityResolutionAlgorithm.cpp:202]: I think the idea here was to keep empty trajectories /cc @paulgessinger

File: Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp
Comment (andiwand) [inline Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:75]:
```suggestion
                      {track->second.first});
```

File: Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp
Comment (andiwand) [inline Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:79]:
```suggestion
      for (const auto& hit : hits) {
        // Add the hits of the new cluster to the hitToTrack
        hitToTrack.emplace(hit, track->second.first);
      }
```

File: Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp
Comment (andiwand) [inline Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:49]:
```suggestion
std::unordered_map<int, std::vector<int>> clusterTracks(
```

File: Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp
Comment (andiwand) [inline Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:133]:
```suggestion
  for (const auto& [key, val] : clusters) {
```

File: Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp
Comment (andiwand) [inline Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:134]:
```suggestion
    for (const auto& track : val) {
```

File: Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp
Comment (andiwand) [inline Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:160]:
```suggestion
  for (const auto & [key, val] : clusters) {
```

File: Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp
Comment (andiwand) [inline Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:163]:
```suggestion
    for (const auto& track : val) {
```

File: Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp
Comment (andiwand) [inline Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:177]:
```suggestion
  for (auto &&iTrack : goodTracks) {
```

File: Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp
Comment (andiwand) [inline Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:178]:
```suggestion
    const auto& outputTips = trackTips.at(iTrack);
```

File: Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_perf.py
Comment (andiwand) [inline Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_perf.py:10]:
```suggestion
```

File: Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_perf.py
Comment (andiwand) [inline Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_perf.py:13]:
```suggestion
```

File: Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_perf.py
Comment (andiwand) [inline Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_perf.py:54]: looks like this function is duplicated in `train_ambiguity_solver.py`. would be cool if we could import it in one of the modules

File: Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_full_chain.py
Comment (andiwand) [inline Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_full_chain.py:66]: :smiling_imp:

File: Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_full_chain.py
Comment (andiwand) [inline Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_full_chain.py:59]: what is the difference to `PrepareDataSet`?

File: Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_full_chain.py
Comment (andiwand) [inline Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_full_chain.py:89]: personally I am not a fan of functions/variables starting with a capital letter. sorry had to complain at least once :smile:

File: Examples/Scripts/Python/MLAmbiguityResolution/train_ambiguity_solver.py
Comment (andiwand) [inline Examples/Scripts/Python/MLAmbiguityResolution/train_ambiguity_solver.py:19]: duplicate?

File: Examples/Scripts/Python/MLAmbiguityResolution/train_ambiguity_solver.py
Comment (andiwand) [inline Examples/Scripts/Python/MLAmbiguityResolution/train_ambiguity_solver.py:61]: duplicate?

File: Examples/Scripts/Python/MLAmbiguityResolution/train_ambiguity_solver.py
Comment (andiwand) [inline Examples/Scripts/Python/MLAmbiguityResolution/train_ambiguity_solver.py:118]: duplicate?

File: Examples/Scripts/Python/MLAmbiguityResolution/train_ambiguity_solver.py
Comment (andiwand) [inline Examples/Scripts/Python/MLAmbiguityResolution/train_ambiguity_solver.py:99]: duplicate?

**Review by andiwand — COMMENTED** [review]
in general the python scripts could share some common code

left a few other comments. nothing critical - as always feel free to ignore

File: Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_full_chain.py
Comment (benjaminhuth) [reply Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_full_chain.py:89]: you have also the official ACTS coding standard on your side :D https://acts.readthedocs.io/en/latest/codeguide.html#n-3-functions-and-methods-use-mixedcase
But not sure how strict we are in this respective

File: Examples/Scripts/Python/full_chain_odd.py
Comment (Corentin-Allaire) [reply Examples/Scripts/Python/full_chain_odd.py:177]: The training uses the csv files as input so yes it is kind of needed. The script I provide for performance evaluation also use Csv so it is also useful to have it for the solver. (They should not be too big ~ 0.1 to 1 Mo per event)

File: Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp
Comment (paulgessinger) [reply Examples/Algorithms/TrackFinding/src/AmbiguityResolutionMLAlgorithm.cpp:187]: I think we should look into porting it to `Acts::TrackContainer` but since I'm working on other parts of the chain still, let's not block this for that.

File: Examples/Algorithms/TrackFinding/src/AmbiguityResolutionAlgorithm.cpp
Comment (Corentin-Allaire) [reply Examples/Algorithms/TrackFinding/src/AmbiguityResolutionAlgorithm.cpp:202]: The reason why I added this is that if we don't and we add the cutoff on the number of hits per track they we have warning in the CI and it fail. I would be happy to change it back though

Comment (paulgessinger) [reply Examples/Algorithms/TrackFinding/src/AmbiguityResolutionAlgorithm.cpp:202]: Not sure I get what you're saying.

File: Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_full_chain.py
Comment (Corentin-Allaire) [reply Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_full_chain.py:66]: That is the only way I found to pass something to a function by reference in python... But we don't need it anymore so I will remove it

File: Examples/Algorithms/TrackFinding/src/AmbiguityResolutionAlgorithm.cpp
Comment (Corentin-Allaire) [reply Examples/Algorithms/TrackFinding/src/AmbiguityResolutionAlgorithm.cpp:202]: Sorry, without `if (!tips.empty()) {` you try to write empty trajectory if `Nmeas<Nminmeas` for a track, which is true many track. Writing a empty track cause a Warning and the CI is configure to fail in case of warning of the full chain. So, now that I have added a cutoff on the minimum number of hit per track before the solver, we would get Warning in the CI, making it fail.

File: Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_full_chain.py
Comment (Corentin-Allaire) [reply Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_full_chain.py:89]: It hurt my brain to not start a line with capital letter but you are 100% right I will go change this

File: Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_perf.py
Comment (Corentin-Allaire) [reply Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_perf.py:54]: Good idea ! I will see what I can do

File: Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_full_chain.py
Comment (Corentin-Allaire) [reply Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_full_chain.py:59]: So prepare `PrepareDataSet` does the reading and formating of the csv input while `PrepareData` prepare the input of the NN (normalisation...). I am really bad at naming stuff very bad name for those function...

File: Examples/Scripts/Python/full_chain_odd.py
Comment (andiwand) [reply Examples/Scripts/Python/full_chain_odd.py:177]: I think we should stick with root and for the training people can just trow in csv directories by hand. because the full chain is just meant for sim+reco I would argue

Comment (Corentin-Allaire) [reply Examples/Scripts/Python/full_chain_odd.py:177]: I don't care that much, but I am not sure I follow your reasoning, why would root be better than Csv for sim+reco ? They are many people that use the Csv files for performance analysis afterward. If the concern is disk space it doesn't really matter as for 100 ttbar event the Csv represent less than 1% of the output.
If you insist on removing it I would argue it should at least be commented out so people know they can change this option.

Comment (andiwand) [reply Examples/Scripts/Python/full_chain_odd.py:177]: I also don't really care but I would like to keep it consistent. so either write everything in root or csv or both.

the only benefit of using root is that you get a single file and the writers tend to be more complete. at least from my experience sometimes the csv output does not contain all the information the root files do

Comment (Corentin-Allaire) [reply Examples/Scripts/Python/full_chain_odd.py:177]: So what I propose is that for all algorithm the let you write to csv I add  `// outputDirCsv=outputDir,` commented out that way it is still consistent and if people try to do stuff that require the Csv they should realise what they need to do to get then csv files without having to ask us directly. Works for you ?

Comment (andiwand) [reply Examples/Scripts/Python/full_chain_odd.py:177]: sounds good to me :+1:

File: Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_perf.py
Comment (Corentin-Allaire) [reply Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_perf.py]: I have done the best I could in that regard, it should be clearer with the new implementation

File: Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_full_chain.py
Comment (Corentin-Allaire) [reply Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_full_chain.py]: Should be done now, I hope I haven't missed one

File: Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_full_chain.py
Comment (Corentin-Allaire) [reply Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_full_chain.py]: The name of the function should be more expicit now.

**Comment by Corentin-Allaire** [comment]
All of the comment should have been addressed now ! I made one tiny change to the file name for the Csv multitrajectory writing, it use to always be `CKFtracks.csv` now the naming depend on the algorithm that called the writer (so it is easier to write the output for the CKF and the solver at the same time) .

File: Examples/Scripts/Python/full_chain_odd.py
Comment (andiwand) [inline Examples/Scripts/Python/full_chain_odd.py:126]:
```suggestion
            outputDirRoot=outputDir,
            # outputDirCsv=outputDir,
```

if you want to include this fix here. otherwise I can do it in a followup pr

**Review by andiwand — APPROVED** [review]
great! thank you!

File: Examples/Scripts/Python/full_chain_odd.py
Comment (Corentin-Allaire) [reply Examples/Scripts/Python/full_chain_odd.py:126]: So I didn't change this because I am not sure it won't break something in the test or not. If you don't mind doing it as a follow up it would speed up the process of getting this merged

Comment (andiwand) [reply Examples/Scripts/Python/full_chain_odd.py:126]: I think this is a debugging leftover from someone

sure no worries

**Review by andiwand — APPROVED** [review]

**Pull request merged**
